EPI/parse-nested-GrFN.py

```diff
@@ -0,0 +1,124 @@
+import json
+from argparse import ArgumentParser
+
+
+def formatGraph(data): 
+    modelMetadata = {}
+    nodesDict = {}
+    nodes = []
+    edges = []
+
+    for key,value in data.items():
+        if key == 'variables':
+            for item in value: 
+                if 'metadata' in item:
+                    metadata = item['metadata'][0]
+                else:
+                    metadata = {}
+                item['nodeType'] = 'variable'
+                item['metadata'] = metadata
+                nodesDict[item['uid']] = item
+        if key == 'functions':
+            for item in value: 
+                if 'metadata' in item:
+                    metadata = item['metadata'][0]
+                else:
+                    metadata = {} 
+                item['nodeType'] = 'function'
+                item['metadata'] = metadata
+                nodesDict[item['uid']] = item
+        if key == 'hyper_edges':
+            for item in value: 
+                if len(item['inputs']) > 0:
+                    for i in item['inputs']: # Inputs array is sometimes empty
+                        first_edge = { 'source': i, 'target': item['function']}
+                        edges.append(first_edge)
+                
+                for o in item['outputs']:
+                    second_edge = { 'source': item['function'], 'target': o }
+                    edges.append(second_edge)
+        if key == 'subgraphs':
+            for item in value: 
+                # Get parent name
+                parent_name = item['scope']
+                if (parent_name == '@global'):
+                    parent_name = 'root'
+
+                # Get container name
+                splitted_id = item['basename'].split('.')
+
+                #Get container metadata
+                if 'metadata' in item:
+                    metadata = item['metadata'][0]
+
+                node = { 'id': item['basename'], 'concept': splitted_id[(len(splitted_id) - 1)], 'nodeType': item['type'], 'label': splitted_id[(len(splitted_id) - 1)], 'parent': parent_name, 'metadata': metadata }
+                nodes.append(node)
+                for n in item['nodes']:
+                    found = nodesDict[n]
+                    # Variables
+                    if (found['nodeType'] == 'variable'):
+                        splitted_identifier = found['identifier'].split('::')
+                        node = { 'id': n, 'concept': splitted_identifier[len(splitted_identifier)-2], 'label': splitted_identifier[len(splitted_identifier)-2], 'nodeType': 'variable', 'type': found['type'], 'parent': item['basename'], 'metadata': found['metadata']}
+                    # Functions
+                    else:
```

Comment:
I know that we have only two node types right now. But--and it's not likely--if in the future we added a third node type, then we would be assuming here that the node would be a function. A more defensive approach would be to check explicitly `elif found['nodeType'] == 'function'`, and then have a final `else` that raises an exception saying something like "unrecognized node type: ...".

Reply:
Yes, that's a good point, I'll include it.